File: examples/unionfind.py
class UnionFind:
    """Union-Find data structure.
    """

    def __init__(self, n: int) -> None:
        """Initialize the Union-Find data structure.
        
        Args:
            n (int): Number of elements.
            
        Returns:
            None
            
        Errors:
            ValueError: If n is less than 1.
        """
        assert n > 0, "n must be positive"
        self.n = n
        self.parents = [-1] * n

    def find(self, i: int) -> int:
        """Find the root of an element.
        
        Args:
            i (int): Element.
            
        Returns:
            int: Root of i.
        """
        if self.parents[i] < 0:
            return i
        else:
            self.parents[i] = self.find(self.parents[i])
            return self.parents[i]

    def union(self, i: int, j: int) -> None:
        """Union two elements.
        
        Args:
            i (int): First element.
            j (int): Second element.
            
        Returns:
            None
        """
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i == root_j:
            return
        if root_i > root_j:
            root_i, root_j = root_j, root_i
        self.parents[root_i] += self.parents[root_j]
        self.parents[root_j] = root_i

    def same(self, i: int, j: int) -> bool:
        """Check if two elements are in the same set.
        
        Args:
            i (int): First element.
            j (int): Second element.
            
        Returns:
            bool: True if i and j are in the same set.
        """
        return self.find(i) == self.find(j)
    
    def size(self, i: int) -> int:
        """Return the size of the set containing i.
        
        Args:
            i (int): Element.
            
        Returns:
            int: Size of the set containing i.
        """
        return -self.parents[self.find(i)]
    
    def max_size(self) -> int:
        """Return the size of the largest set.
        
        Returns:
            int: Size of the largest set.
        """
        return -min(self.parents)
    
    def __repr__(self) -> str:
        """Return a string representation of the Union-Find data structure.
        
        Returns:
            str: String representation of the Union-Find data structure.
        """
        return f"UnionFind(n = {self.n})"

File: src/unionfind.rs
use std::mem::swap;

pub struct UnionFind {
    // n: number of elements
    // parents: parent of each element
    n: usize,
    parents: Vec<i64>,
}

impl UnionFind{
    pub fn new(n: usize) -> UnionFind {
        // initialize UnionFind
        UnionFind {
            n: n,
            parents: vec![-1; n],
        }
    }

    pub fn find(&mut self, i: usize) -> usize {
        // find root of i
        if self.parents[i] < 0 {
            return i;
        } else {
            let parent_i = self.parents[i] as usize;
            self.parents[i] = self.find(parent_i) as i64;
            return self.parents[i] as usize;
        }
    }

    pub fn union(&mut self, i: usize, j: usize) -> () {
        // unite i and j
        let mut root_i = self.find(i);
        let mut root_j = self.find(j);
        if root_i == root_j {
            return;
        }
        if root_i > root_j {
            swap(&mut root_i, &mut root_j);
        }
        self.parents[root_i] += self.parents[root_j];
        self.parents[root_j] = root_i as i64;
    }

    pub fn same(&mut self, i: usize, j: usize) -> bool {
        // check if i and j are in the same group
        return self.find(i) == self.find(j);
    }

    pub fn size(&mut self, i: usize) -> i64 {
        // return size of group of i
        let root_i = self.find(i);
        return -self.parents[root_i];
    }

    pub fn max_size(&mut self) -> i64 {
        // return max size of groups
        let mut max_val: i64 = 0;
        for i in 0..self.n {
            if max_val < -self.parents[i] {
                max_val = -self.parents[i];
            }
        }
        return max_val;
    }

}